include/nbl/asset/utils/CVertexHashGrid.h

```diff
@@ -0,0 +1,212 @@
+#ifndef _NBL_ASSET_C_VERTEX_HASH_MAP_H_INCLUDED_
+#define _NBL_ASSET_C_VERTEX_HASH_MAP_H_INCLUDED_
+
+#include "nbl/core/declarations.h"
+
+namespace nbl::asset
+{
+
+template <typename T>
+concept HashGridVertexData = requires(T obj, T const cobj, uint32_t hash) {
+		{ cobj.getHash() } -> std::same_as<uint32_t>;
+		{ obj.setHash(hash) } -> std::same_as<void>;
+		{ cobj.getPosition() } -> std::same_as<hlsl::float32_t3>;
+};
+
+template <HashGridVertexData VertexData>
+class CVertexHashGrid
+{
+public:
+
+	using vertex_data_t = VertexData;
+	using collection_t = core::vector<VertexData>;
+	struct BucketBounds
+	{
+		collection_t::const_iterator begin;
+		collection_t::const_iterator end;
+	};
+
+	CVertexHashGrid(size_t _vertexCount, uint32_t _hashTableMaxSize, float _cellSize) :
+		m_sorter(createSorter(_vertexCount)),
+		m_hashTableMaxSize(_hashTableMaxSize),
+		m_cellSize(_cellSize)
+	{
+		assert((core::isPoT(m_hashTableMaxSize)));
+
+		m_vertices.reserve(_vertexCount);
+	}
+
+	//inserts vertex into hash table
+	void add(VertexData&& vertex)
+	{
+		vertex.setHash(hash(vertex));
+		m_vertices.push_back(vertex);
+	}
+
+	void validate()
+	{
+		const auto oldSize = m_vertices.size();
+		m_vertices.resize(oldSize*2u);
+		auto finalSortedOutput = std::visit( [&](auto& sorter) { return sorter(m_vertices.data(), m_vertices.data() + oldSize, oldSize, KeyAccessor()); },m_sorter );
+
+		if (finalSortedOutput != m_vertices.data())
+			m_vertices.erase(m_vertices.begin(), m_vertices.begin() + oldSize);
+		else
+			m_vertices.resize(oldSize);
+	}
+
+	const collection_t& vertices() const { return m_vertices; }
+
+	collection_t& vertices(){ return m_vertices; }
+
+	inline uint32_t getVertexCount() const { return m_vertices.size(); }
+
+	template <typename Fn>
+	void iterateBroadphaseCandidates(const VertexData& vertex, Fn fn) const
+	{
+		std::array<uint32_t, 8> neighboringCells;
+		const auto cellCount = getNeighboringCellHashes(neighboringCells.data(), vertex);
+
+		//iterate among all neighboring cells
+		for (uint8_t i = 0; i < cellCount; i++)
+		{
+			const auto& neighborCell = neighboringCells[i];
+			BucketBounds bounds = getBucketBoundsByHash(neighborCell);
+			for (; bounds.begin != bounds.end; bounds.begin++)
+			{
+				const vertex_data_t& neighborVertex = *bounds.begin;
+				if (&vertex != &neighborVertex)
+					if (!fn(neighborVertex)) break;
```

Comment:
add some documentation in a comment that `Fn` needs to return boolean (I think its actually checkable with a cocnept that the signature of `operator()` matches nicely) and it should always return `true` unless it wishes to break out of the iteration